### PR TITLE
Adding ability for spackbot to re-run a GitLab CI pipeline

### PR DESCRIPTION
**Pull request by vsoch**
This PR will add support for asking spackbot to re-run a pipeline! I tested this with cancelled pipelines and the triggers worked. It works by way of checking if the person has write, and if so, allowing them to trigger the endpoint to run the GitLab branch.

I duplicated helpers.py since I needed the found function, which is in the other PR.

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

### CODE REVIEW

File: spackbot/pr_run_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+# We can only make the request with a GITLAB TOKEN
+GITLAB_TOKEN = os.environ.get("GITLAB_TOKEN")
+
+
+async def run_pipeline(event, gh):
+    """
+    Make a request to re-run a pipeline.
+    """
+    # Get the pull request number
+    number = event.data["issue"]["pull_request"]["url"].split("/")[-1]
```
Comment (tgamblin) [inline spackbot/pr_run_pipeline.py:28]: isn't the PR number already an int in `event.data["issue"]["pull_request"]["number"]`?

File: spackbot/pr_run_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+    # We need the pull request branch
+    response = requests.get(event.data["issue"]["pull_request"]["url"])
```
Comment (tgamblin) [inline spackbot/pr_run_pipeline.py:31]: Seems like `event.data["issue"]["pull_request"]` could just be pulled out initially and used in both of these expressions.

File: spackbot/pr_run_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            "Sorry %s, I cannot do that for you. Only users with write can make this request!"
+            % sender
+        )
+
+    # We need the branch name plus number to assemble the GitLab CI
+    branch = pr["head"]["ref"]
+    branch = "github/pr%s_%s" % (number, branch)
+
+    # Spack has project ID 2
+    url = "https://gitlab.spack.io/api/v4/projects/2/pipeline?ref=%s" % branch
```
Comment (tgamblin) [inline spackbot/pr_run_pipeline.py:52]: can you make this part a global constant like:

```python
gitlab_spack_project_url = "https://gitlab.spack.io/api/v4/projects/2"
```

File: spackbot/pr_run_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    headers = {"PRIVATE-TOKEN": GITLAB_TOKEN}
+    response = requests.post(url, headers=headers)
+    result = response.json()
+    if "detailed_status" in result and "details_path" in result["detailed_status"]:
+        url = "https://gitlab.spack.io/%s" % result["detailed_status"]["details_path"]
```
Comment (tgamblin) [inline spackbot/pr_run_pipeline.py:57]: `"gitlab.spack.io"` should probably also be a constant up top

File: spackbot/pr_run_pipeline.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        return
+
+    # spackbot should not respond to himself!
+    if re.search(alias_regex, event.data["comment"]["user"]["login"]):
+        return
+
+    # Respond with appropriate messages
+    comment = event.data["comment"]["body"]
+
+    # @spackbot hello
```
Comment (tgamblin) [inline spackbot/pr_run_pipeline.py:77]: is this copy/paste?

**Review by tgamblin — CHANGES_REQUESTED** [review]
@vsoch LGTM! minor stuff:

File: spackbot/pr_run_pipeline.py (hunk 1 above)
Comment (vsoch) [reply spackbot/pr_run_pipeline.py:28]: Nope. But you can get it from the end of the url there.

```
    "pull_request": {
      "url": "https://api.github.com/repos/vsoch/spack/pulls/15",
      "html_url": "https://github.com/vsoch/spack/pull/15",
      "diff_url": "https://github.com/vsoch/spack/pull/15.diff",
      "patch_url": "https://github.com/vsoch/spack/pull/15.patch"
    },
```

File: spackbot/pr_run_pipeline.py (hunk 2 above)
Comment (vsoch) [reply spackbot/pr_run_pipeline.py:31]: sure thing.

File: spackbot/pr_run_pipeline.py (hunk 3 above)
Comment (vsoch) [reply spackbot/pr_run_pipeline.py:52]: sure.

File: spackbot/pr_run_pipeline.py (hunk 4 above)
Comment (vsoch) [reply spackbot/pr_run_pipeline.py:57]: sure

File: spackbot/pr_run_pipeline.py (hunk 5 above)
Comment (vsoch) [reply spackbot/pr_run_pipeline.py:77]: Yes forgot to update the comment!

File: spackbot/pr_run_pipeline.py (hunk 1 above)
Comment (vsoch) [reply spackbot/pr_run_pipeline.py]: I think the idea is that the developer is expected to use the PR url to get metadata about the PR. But we just need the number so it's easier to just grab it from the url.

Comment (tgamblin) [reply spackbot/pr_run_pipeline.py]: Yeah I agree this is better.

**Review by tgamblin — APPROVED** [review]

**Pull request merged**